File: app/src/main/java/com/fullmeadalchemist/mustwatch/ui/batch/form/BatchFormViewModel.java
/*
 * Copyright (c) 2017 Full Mead Alchemist, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.fullmeadalchemist.mustwatch.ui.batch.form;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;

import com.fullmeadalchemist.mustwatch.repository.BatchIngredientRepository;
import com.fullmeadalchemist.mustwatch.repository.BatchRepository;
import com.fullmeadalchemist.mustwatch.repository.IngredientRepository;
import com.fullmeadalchemist.mustwatch.repository.RecipeRepository;
import com.fullmeadalchemist.mustwatch.repository.UserRepository;
import com.fullmeadalchemist.mustwatch.vo.Batch;
import com.fullmeadalchemist.mustwatch.vo.BatchIngredient;
import com.fullmeadalchemist.mustwatch.vo.Ingredient;
import com.fullmeadalchemist.mustwatch.vo.Recipe;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;


public class BatchFormViewModel extends ViewModel {

    protected Batch batch;
    private BatchRepository batchRepository;
    private BatchIngredientRepository batchIngredientRepository;
    private RecipeRepository recipeRepository;
    private IngredientRepository ingredientRepository;
    private UserRepository userRepository;

    @Inject
    public BatchFormViewModel(BatchRepository batchRepository, BatchIngredientRepository batchIngredientRepository, RecipeRepository recipeRepository, IngredientRepository ingredientRepository, UserRepository userRepository) {
        this.batchRepository = batchRepository;
        this.batchIngredientRepository = batchIngredientRepository;
        this.recipeRepository = recipeRepository;
        this.ingredientRepository = ingredientRepository;
        this.userRepository = userRepository;
    }

    public LiveData<Batch> getBatch(Long id) {
        return batchRepository.getBatch(id);
    }

    public void addBatch(Batch batch) {
        if (batch.ingredients == null) {
            batch.ingredients = new ArrayList<>();
        }
        batchRepository.addBatch(batch);
    }

    void addIngredient(BatchIngredient batchIngredient) {
        if (batch.ingredients == null){
            batch.ingredients = new ArrayList<>();
        }
        batch.ingredients.add(batchIngredient);
    }

    public LiveData<Long> getCurrentUserId() {
        return userRepository.getCurrentUserId();
    }

    public void updateBatch() {
        if (batch.ingredients == null){
            batch.ingredients = new ArrayList<>();
        }
        batchRepository.updateBatch(batch);
    }

    public LiveData<Long> saveNewBatch() {
        return batchRepository.addBatch(batch);
    }

    public LiveData<Recipe> getRecipe(long recipeId) {
        return recipeRepository.getRecipe(recipeId);
    }

    public LiveData<Ingredient> getIngredient(String id) {
        return ingredientRepository.getIngredientById(id);
    }

    public LiveData<List<Ingredient>> getSugars() {
        return ingredientRepository.getSugarEntries();
    }

    public LiveData<List<Ingredient>> getNutrients() {
        return ingredientRepository.getNutrientEntries();
    }

    public LiveData<List<Ingredient>> getYeasts() {
        return ingredientRepository.getYeastEntries();
    }

    public LiveData<List<Ingredient>> getStabilizers() {
        return ingredientRepository.getStabilizerEntries();
    }

    public LiveData<List<BatchIngredient>> getBatchIngredients(long batchId) {
        return batchRepository.getBatchIngredients(batchId);
    }

    public LiveData<List<Recipe>> getRecipes(Long userId) {
        if (userId == null) {
            return recipeRepository.getPublicRecipes();
        } else {
            return recipeRepository.getRecipes(userId);
        }
    }

    public LiveData<List<BatchIngredient>> getRecipeIngredients(long recipeId) {
        return recipeRepository.getRecipeIngredients(recipeId);
    }
}

File: app/src/main/java/com/fullmeadalchemist/mustwatch/ui/batch/form/BatchFormViewModel.kt
/*
 * Copyright (c) 2017 Full Mead Alchemist, LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.fullmeadalchemist.mustwatch.ui.batch.form

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.fullmeadalchemist.mustwatch.repository.*
import com.fullmeadalchemist.mustwatch.vo.*
import java.util.*

class BatchFormViewModel(
        private val batchRepository: BatchRepository,
        private val batchIngredientRepository: BatchIngredientRepository,
        private val recipeRepository: RecipeRepository,
        private val ingredientRepository: IngredientRepository,
        private val userRepository: UserRepository
) : ViewModel() {

    var batch = MutableLiveData<Batch>()

    val currentUserId: LiveData<User>
        get() = userRepository.getCurrentUser()

    val sugars: LiveData<List<Ingredient>>
        get() = ingredientRepository.getSugars()

    val nutrients: LiveData<List<Ingredient>>
        get() = ingredientRepository.getNutrients()

    val yeasts: LiveData<List<Ingredient>>
        get() = ingredientRepository.getYeasts()

    val stabilizers: LiveData<List<Ingredient>>
        get() = ingredientRepository.getStabilizers()

    fun getBatch(id: Long?): LiveData<Batch> {
        return batchRepository.getBatch(id)
    }

    fun addBatch(batch: Batch) {
        if (batch.ingredients == null) {
            batch.ingredients = arrayListOf()
        }
        batchRepository.addBatch(batch)
    }

    internal fun addIngredient(batchIngredient: BatchIngredient) {
        batch.value?.let { b ->
            if (b.ingredients == null) {
                b.ingredients = ArrayList()
            }
            (b.ingredients as ArrayList).let { bi ->
                bi.add(batchIngredient)
                batchIngredientRepository.addBatchIngredients(bi)
            }
        }
    }

    fun updateBatch(batch: Batch) {
        if (batch.ingredients == null) {
            batch.ingredients = ArrayList()
        }
        batchRepository.updateBatch(batch)
    }

    fun saveNewBatch(batch: Batch): LiveData<Long> {
        return batchRepository.addBatch(batch)
    }

    fun getRecipe(recipeId: Long): LiveData<Recipe> {
        return recipeRepository.getRecipe(recipeId)
    }

    fun getIngredient(id: String): LiveData<Ingredient> {
        return ingredientRepository.getIngredientById(id)
    }

    fun getBatchIngredients(batchId: Long): LiveData<List<BatchIngredient>> {
        return batchRepository.getBatchIngredients(batchId)
    }

    fun getRecipes(userId: UUID?): LiveData<List<Recipe>> {
        return if (userId == null) {
            recipeRepository.getPublicRecipes()
        } else {
            recipeRepository.getRecipes(userId)
        }
    }

    fun getRecipeIngredients(recipeId: Long): LiveData<List<BatchIngredient>> {
        return recipeRepository.getRecipeIngredients(recipeId)
    }
}
